consensus/obcpbft/events.go

```diff
@@ -0,0 +1,306 @@
+	stopChan  chan struct{}    // Channel to deliver the timer stop events to the service go routine
+	manager   eventManager     // The event manager to deliver the event to after timer expiration
+}
+
+// newEventTimer creates a new instance of eventTimerImpl
+func newEventTimer(manager eventManager) eventTimer {
+	et := &eventTimerImpl{
+		startChan: make(chan *timerStart),
+		stopChan:  make(chan struct{}),
+		threaded:  threaded{make(chan struct{})},
+		manager:   manager,
+	}
+	go et.loop()
+	return et
+}
+
+// softReset tells the timer to start a new countdown, only if it is not currently counting down
+// this will not clear any pending events
+func (et *eventTimerImpl) softReset(timeout time.Duration, event event) {
+	et.startChan <- &timerStart{
+		duration: timeout,
+		event:    event,
+		hard:     true,
+	}
+}
+
+// reset tells the timer to start counting down from a new timeout, this also clears any pending events
+func (et *eventTimerImpl) reset(timeout time.Duration, event event) {
+	et.startChan <- &timerStart{
+		duration: timeout,
+		event:    event,
+		hard:     false,
+	}
+}
+
+// stop tells the timer to stop, and not to deliver any pending events
+func (et *eventTimerImpl) stop() {
+	et.stopChan <- struct{}{}
+}
+
+// loop is where the timer thread lives, looping
+func (et *eventTimerImpl) loop() {
+	var eventDestChan chan<- event
+	var event event
+
+	for {
+		// A little state machine, relying on the fact that nil channels will block on read/write indefinitely
+
+		select {
+		case start := <-et.startChan:
+			if et.timerChan != nil {
+				if start.hard {
+					logger.Debug("Resetting a running timer")
+				} else {
+					continue
+				}
+			}
+			logger.Debug("Starting timer")
+			et.timerChan = time.After(start.duration)
+			if eventDestChan != nil {
+				logger.Debug("Timer cleared pending event")
+			}
+			event = start.event
+			eventDestChan = nil
+		case <-et.stopChan:
+			if et.timerChan == nil && eventDestChan == nil {
+				logger.Warning("Attempting to stop an unfired idle timer")
+			}
+			et.timerChan = nil
+			logger.Debug("Stopping timer for")
+			if eventDestChan != nil {
+				logger.Debug("Timer for cleared pending event")
+			}
+			eventDestChan = nil
+			event = nil
+		case <-et.timerChan:
+			logger.Debug("Event timer fired")
+			et.timerChan = nil
+			eventDestChan = et.manager.queue()
+		case eventDestChan <- event:
+			logger.Debug("Timer event delivered")
+			eventDestChan = nil
+		case <-et.exit:
+			logger.Debug("Halting timer")
+			return
+		}
+	}
+}
+
+// ------------------------------------------------------------
+//
+// Event Types
+//
+// ------------------------------------------------------------
+
+type eventType int
+
+const (
+	workEventID eventType = iota
+	viewChangeTimerEventID
+	execDoneEventID
+	stateUpdatedEventID
+	stateUpdatingEventID
+	messageEventID
+)
+
+// workEvent is a temporary type, to inject work
+type workEvent func()
+
+func (e workEvent) eventType() eventType {
```

Comment:
Yes, I agree, I originally implemented this via type switching, but ran into a problem that the const struct helped addressed, I'll be moving it back to type switching.